### PR TITLE
[WIP] Support LR_ReduceOnPlateau

### PR DESCRIPTION
**Pull request by gengenkai**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification
Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmcv/runner/hooks/lr_updater.py [hunk 1]
```diff
@@ -614,3 +614,184 @@ def format_param(name, optim, param):
         if name not in param:
             raise KeyError(f'{name} is not found in {param.keys()}')
         return param[name]
+
+
+@HOOKS.register_module()
+class ReduceLrUpdateHook(LrUpdaterHook):
+
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:621]: please add doctring

File: mmcv/runner/hooks/lr_updater.py [hunk 2]
```diff
@@ -614,3 +614,184 @@ def format_param(name, optim, param):
...
+                 factor=0.1,
+                 patience=10,
+                 threshold=1e-4,
+                 threshold_mode='rel',
+                 cooldown=0,
+                 min_lr=0.,
+                 eps=1e-8,
+                 **kwargs):
+        if isinstance(periods, list):
+            assert mmcv.is_list_of(periods, int)
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:635]: should add arguments like `assert expression [, arguments]`

File: mmcv/runner/hooks/lr_updater.py [hunk 3]
```diff
@@ -614,3 +614,184 @@ def format_param(name, optim, param):
         if name not in param:
             raise KeyError(f'{name} is not found in {param.keys()}')
         return param[name]
+
+
+@HOOKS.register_module()
+class ReduceLrUpdateHook(LrUpdaterHook):
+
+    def __init__(self,
+                 periods,
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:623]: add type hint

File: mmcv/runner/hooks/lr_updater.py [hunk 4]
```diff
@@ -614,3 +614,184 @@ def format_param(name, optim, param):
...
+                 eps=1e-8,
+                 **kwargs):
+        if isinstance(periods, list):
+            assert mmcv.is_list_of(periods, int)
+            assert all([s > 0 for s in periods])
+        else:
+            raise TypeError('"periods" must be a list')
+        self.periods = periods
+        self.val_metric = val_metric
+        if mode not in ['min', 'max']:
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:641]: it is readable to add a blank line between different logical code

File: mmcv/runner/hooks/lr_updater.py [hunk 5]
```diff
@@ -614,3 +614,184 @@ def format_param(name, optim, param):
...
+            rel_epsilon = 1. - self.threshold
+            return a < best * rel_epsilon
+
+        elif self.mode == 'min' and self.threshold_mode == 'abs':
+            return a < best - self.threshold
+
+        elif self.mode == 'max' and self.threshold_mode == 'rel':
+            rel_epsilon = 1. + self.threshold
+            return a > best * rel_epsilon
+
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:717]: remove the blank line

File: mmcv/runner/hooks/lr_updater.py [hunk 6]
```diff
@@ -614,3 +614,221 @@ def format_param(name, optim, param):
...
+    """ReduceLROnPlateau Scheduler.
+
+    Reduce learning rate when a metric has stopped improving. This scheduler
+    reads a metrics quantity and if no improvement is seen for a 'patience'
+    number of epochs, the learning rate is reduced.
+
+    Args:
+        periods (list[int]): Periods that taking the metric value in count.
+        val_metric (string): The key of the validation metric in outputs. If
+        val_metric is None, the metrics will be loss value.
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:630]: lack of indent

File: mmcv/runner/hooks/lr_updater.py [hunk 7]
```diff
@@ -614,3 +614,221 @@ def format_param(name, optim, param):
...
+                 threshold_mode='rel',
+                 cooldown=0,
+                 min_lr=0.,
+                 eps=1e-8,
+                 **kwargs):
+        if isinstance(periods, list):
+            assert mmcv.is_list_of(periods, int)
+            assert all([s >= 0 for s in periods])
+        else:
+            raise TypeError('"periods" must be a list')
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:676]: Line #L676 was not covered by tests.

File: mmcv/runner/hooks/lr_updater.py [hunk 8]
```diff
@@ -614,3 +614,221 @@ def format_param(name, optim, param):
...
+            raise ValueError(
+                'mode must be one of "min" or "max", instead got {mode}')
+        self.mode = mode
+        if factor >= 1.0:
+            raise ValueError('Factor should be < 1.0')
+        self.factor = factor
+        self.patience = patience
+        self.threshold = threshold
+        if threshold_mode not in ['rel', 'abs']:
+            raise ValueError('thresh_mode must be one of "rel" or "abs",\
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:689]: Line #L689 was not covered by tests

File: mmcv/runner/hooks/lr_updater.py [hunk 9]
```diff
@@ -614,3 +615,223 @@ def format_param(name, optim, param):
...
             raise KeyError(f'{name} is not found in {param.keys()}')
         return param[name]
+
+
+@HOOKS.register_module()
+class ReduceLrUpdateHook(LrUpdaterHook):
+    """ReduceLROnPlateau Scheduler.
+
+    Reduce learning rate when a metric has stopped improving. This scheduler
+    reads a metrics quantity and if no improvement is seen for a 'patience'
```
Comment (ZwwWayne) [inline mmcv/runner/hooks/lr_updater.py:625]: and the learning rate will be reduced if no improvement is observed for a ``patience`` number of epochs.

**Comment by zhouzaida** [comment]
@gengenkai ,please fix the CI

**Pull request closed without merging**